### PR TITLE
Dop 50 terraform GitHub repository module

### PR DESCRIPTION
**Pull request by tudormangri10**

### CODE REVIEW

**Comment by mihneastaub** [comment]
LGTM

Please add the followings to this repository:
- terraform pre-commit without the tflint ([docs](https://github.com/coremaker/devops-docs/blob/main/repositories/terraform_pre_commit.md))
- pull request template (https://github.com/coremaker/devops-docs/blob/main/repositories/pull_request_templates.md)

Also, it will be very useful if you will create an example directory with the usage of the module. 
Something like [this](https://github.com/terraform-aws-modules/terraform-aws-eks/tree/master/examples/complete). In this example directory, please include also the providers and everything.

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,61 @@
+
```
Comment (mihneastaub) [inline main.tf]: Please format the code, you can use this command `terraform fmt` in the root directory.

File: README.md [hunk 2]
```diff
@@ -1 +1,56 @@
...
+|------|------|
+| [github_repository](https://registry.terraform.io/providers/integrations/github/latest/docs/resources/repository) | resource |
+| [github_team_repository](https://registry.terraform.io/providers/integrations/github/latest/docs/resources/team_repository) | resource |
+| [github_branch_protection](https://registry.terraform.io/providers/integrations/github/latest/docs/resources/branch_protection) | resource |
+| [github_repository_deploy_key](https://registry.terraform.io/providers/integrations/github/latest/docs/resources/repository_deploy_key) | resource |
+
+# Usage Example
+```terraform
+module "github-repository" {
+source = "github.com/coremaker/terraform-github-repository.git?ref=DOP-50-terraform-github-repository-Module"
```
Comment (CosminBriscaru) [inline README.md]: use root module for examples without a tagged version

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Repository configuration module
+resource "github_repository" "coremaker_repo" {
```
Comment (CosminBriscaru) [inline main.tf]: let's use a more common and meaningful name for all resources.

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      required_approving_review_count = required_pull_request_reviews.value.required_approving_review_count
+    }
+  }
+}
+
+# Deploy keys for repository auth
+resource "github_repository_deploy_key" "example_repository_deploy_key" {
+  title      = "Repository keys"
+  repository = github_repository.coremaker_repo.name
+  for_each   = toset(var.deploy_key)
```
Comment (CosminBriscaru) [inline main.tf:54]: It would be better for visibility that the for_each arguments to be first.

File: README.md [hunk 5]
```diff
@@ -1 +1,56 @@
...
+        }]
+        required_pull_request_reviews = [{
+          dismiss_stale_reviews             = true  # Dismiss approved reviews automatically when a new commit is pushed. Defaults to false
+          require_code_owner_reviews        = false # Require an approved review in pull requests including files with a designated code owner. Defaults to false.
+          required_approving_review_count   = "2"   # Require x number of approvals to satisfy branch protection requirements
+        }]
+    }
+  }
+  
+  deploy_key = []
```
Comment (CosminBriscaru) [inline README.md]: let's add an example here as well.

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  description = var.repo_description
+  auto_init   = true
+
+  visibility = var.repo_visibility
+}
+
+# Add teams to the given repository
+resource "github_team_repository" "coremaker_team_repo" {
+  repository = github_repository.coremaker_repo.name
+  count      = length(var.teams)
```
Comment (CosminBriscaru) [inline main.tf]: For better visibility usually count and for each arguments are first.

File: variables.tf [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      strict = bool # Require branches to be up to date before merging. Defaults to false
+    }))
+    required_pull_request_reviews = list(object({
+      dismiss_stale_reviews           = bool   # Dismiss approved reviews automatically when a new commit is pushed. Defaults to false
+      require_code_owner_reviews      = bool   # Require an approved review in pull requests including files with a designated code owner. Defaults to false.
+      required_approving_review_count = number # Require x number of approvals to satisfy branch protection requirements
+    }))
+
+  }))
+  description = "Branch protections for selected branches in repository"
```
Comment (CosminBriscaru) [inline variables.tf]: This should be configurable

File: variables.tf [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+variable "repo_description" {
+  default = ""
+}
+
+variable "repo_visibility" {
+  default = ""
+}
+
+variable "teams" {
```
Comment (CosminBriscaru) [inline variables.tf]: lets follow the other names and use repo_teams here as well

File: variables.tf [hunk 9]
```diff
@@ -0,0 +1,47 @@
+variable "repo_name" {
```
Comment (CosminBriscaru) [inline variables.tf:1]: all vars should have a type

**Review by CosminBriscaru — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 8 above)
Comment (tudormangri10) [reply variables.tf]: Done

File: main.tf (hunk 6 above)
Comment (tudormangri10) [reply main.tf]: Done

File: main.tf (hunk 4 above)
Comment (tudormangri10) [reply main.tf:54]: Done

File: main.tf (hunk 1 above)
Comment (tudormangri10) [reply main.tf]: Done

File: variables.tf (hunk 9 above)
Comment (tudormangri10) [reply variables.tf:1]: Done

File: main.tf (hunk 3 above)
Comment (tudormangri10) [reply main.tf]: Done

**Review by mihneastaub — APPROVED** [review]
LGTM

**Pull request merged**